File: 백준/Silver/15666. N과 M （12）/N과 M （12）.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())

nums = list(map(int, input().split()))
visited = [0 for _ in range(n)]
nums.sort()

temp = []
ans = set()


def dfs(start):
    if len(temp) == m:
        print(*temp)
        return
    overlap = 0
    for i in range(start, n):
        if overlap != nums[i]:
            visited[i] = 1
            temp.append(nums[i])
            overlap = nums[i]
            dfs(i)
            temp.pop()
            visited[i] = 0


dfs(0)

File: 백준/Silver/15666. N과 M （12）/N과 M （12）.java
import java.io.*;
import java.util.*;

public class Main {

    static Stack<Integer> stack;
    static boolean[] visited;
    static Set<List<Integer>> set;
    static int n, m;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        stack = new Stack<>();
        set = new LinkedHashSet<>();
        arr = new int[n];
        visited = new boolean[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(arr);

        dfs(0);

        StringBuilder sb = new StringBuilder();
        for (List<Integer> x : set) {
            for (int a : x) {
                sb.append(a).append(" ");
            }
            sb.append("\n");

        }

        bw.write(sb.toString());

        bw.flush();
        bw.close();
        br.close();
    }

    static void dfs(int idx) {
        if (stack.size() == m) {
            set.add(new ArrayList<>(stack));
            return;
        }

        for (int i = idx; i < n; i++) {
            stack.push(arr[i]);
            dfs(i);
            stack.pop();
        }
    }
}